### PR TITLE
cpu: Add PIC32MZ support

### PR DESCRIPTION
**Pull request by neiljay**
~~This PR is dependant on #6060~~ - merged

Add support for PIC32 MZ devices which use a MIPS M5150 core.

This PR adds basic support for Microchip PIC32MZ devices, including
bootstrapping the device, the build system produces a hex image loadable via
MPLAB-IPE. It has been tested on a Digilent Wifire board.

NOTE: This port requires the use of the GCC based MIPS Codescape SDK toolchain
available here:

https://community.imgtec.com/developers/mips/tools/codescape-mips-sdk/download-codescape-mips-sdk-essentials/

You cannot use Microchip's MPLAB / Harmony tools to build this port.

The port has been tested with the following examples:
hello-world
ipc_pingpong
timer_periodic_wakeup
riot_and_cpp

Note this is a re-work of #5885 with review comments addressed but now broken down into 3 separate PR's of which this is the second.

### CODE REVIEW

**Comment by OlegHahm** [comment]
I was able to build and flash the ipc_pingpong (at least mdb is saying:
```
Programming target...
Program succeeded.
```
but I couldn't get any output.

**Comment by OlegHahm** [comment]
Do I assume correctly that UART0 is configured over the mini-USB port?

**Comment by neiljay** [comment]
Its UART4 on the pic device, but yes the MINI USB header is the one you
want, 9600 8N1.

If you want to power the board off that USB cable you need to set the
'VUSelect' Jumper to 'UART'.

you might need to press enter a couple of times in your console to wake
up the USB<->UART converter.

Neill

**Comment by OlegHahm** [comment]
Ah, true. Forgot about the 9600 baud... Now it is working! 👍

**Comment by neiljay** [comment]
Sweet!

**Comment by neiljay** [comment]
Hi all,

Why are we running cppcheck on .s and .S files ?

I get this error:
`
cpu/mips_pic32_common/reset_mod.S:84: error (syntaxError): syntax error
`
from this line:
`
beqz  s1, init_resources	  # Branch if this is NOT an NMI exception.
`
The code assembles fine, I tried an unnamed label '1f' but it didn't like that either...

It looks like this check fails on other (in tree) asm file also (testing in my local docker image):

cpu/lpc2387/asmfunc.s:0: error (syntaxError): syntax error

Hopefully this is now close to a point where we can merge it, The PIC32 ports should be of more interest as the targets are readily available and cheap. @OlegHahm Can you try this on your board.

I ran the unittest suite and that all passed:

 main(): This is RIOT! (Version: 2017.04-devel-152-g204aa-a2f4afe6403a-pr/add_mips-pic32_mz)
........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
OK (712 tests)

**Comment by neiljay** [comment]
So cppcheck was failing on the asm style comments (in an asm file) WTF!!

**Comment by gebart** [comment]
I'm seeing errors with objcopy at the end:
```
   text	   data	    bss	    dec	    hex	filename
  78616	   2424	   6012	  87052	  1540c	/home/jgn/work/src/riot/examples/default/bin/pic32-clicker/default.elf
objcopy: /home/jgn/work/src/riot/examples/default/bin/pic32-clicker/default.hex: address 0xffffffff1d000000 out of range for Intel Hex file
objcopy:/home/jgn/work/src/riot/examples/default/bin/pic32-clicker/default.hex: Bad value
objcopy: --change-section-lma .gcc_except_table+0xffffffff80000000 never used
make: *** [/home/jgn/work/src/riot/examples/default/../../Makefile.include:278: all] Error 1
```

Both pic32-wifire and pic32-clicker.

GNU objcopy (Gentoo 2.27 p1.0) 2.27

File: cpu/mips_pic32mz/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+				--change-section-lma .fini-0x80000000 \
+				--change-section-lma .eh_frame-0x80000000 \
+				--change-section-lma .gcc_except_table-0x80000000 \
+				--change-section-lma .jcr-0x80000000 \
+				--change-section-lma .ctors-0x80000000 \
+				--change-section-lma .dtors-0x80000000 \
+				--change-section-lma .rodata-0x80000000 \
+				--change-section-lma .data-0x80000000 \
+				--change-section-lma .bss-0x80000000 \
+				--change-section-lma .startdata-0x80000000
```
Comment (gebart) [inline cpu/mips_pic32mz/Makefile.include]: tabs in indent, use spaces instead, and I think 4 spaces should be enough here.

**Comment by neiljay** [comment]
Are you sure that its not calling the MIPS toolchain objcopy (which has
this very same bug) ?

I added this to the makefiles:

export OBJCOPY = objcopy #use system objcopy as toolchain one is broken.

There was an issue originally with makefile include order which made
this change necessary too (but the last refactor should have fixed this):
https://github.com/RIOT-OS/RIOT/pull/6639/commits/d273ecd36ba5ed5e61bb4bcf4fa4c1e9aa17c8e6

**Comment by gebart** [comment]
Yeah, it's calling the native objcopy in Gentoo. In docker (riot/riotbuild image, Ubuntu) it works however. I wonder if there's a flag missing in the command to ensure that those address changes are performed correctly. It seems like there may be some confusion between 32 bit and 64 bit (0xffffffff80000000?)

**Comment by neiljay** [comment]
No its a genuine bug in objcopy.

**Comment by gebart** [comment]
@neiljay could you add a reference to any upstream bug reports regarding the bug, it will make it easier to see in the future when the problem has been fixed.

File: cpu/mips_pic32mz/Makefile.include (hunk 1 above)
Comment (neiljay) [inline cpu/mips_pic32mz/Makefile.include]: done.

**Comment by neiljay** [comment]
@gebart 

I'm now getting failures with the unit test build (building relic) which look related to vfs.

```
In file included from ../../../sys/include/vfs.h:69:0,
                 from ../../../sys/include/fs/devfs.h:29,
                 from tests-devfs.c:22:
../../../sys/posix/include/sys/statvfs.h:50:5: error: unknown type name 'fsblkcnt_t'
     fsblkcnt_t f_blocks;     /**< Total number of blocks on file system in

```

Now #5616 is merged are we going to have to merge #6639 (rebased) instead of this and #6092  ?

**Comment by gebart** [comment]
I would be OK if you just do the same ifdef workaround in statvfs.h and fcntl.h that I added for BOARD_MIPS_MALTA in the VFS PR. Then we remove the workaround in the Makefile cleanup PR.

**Comment by neiljay** [comment]
@gebart Are you happy for me to squash this now so we can get this merged ?

File: boards/pic32-wifire/Makefile.features [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Put defined MCU peripherals here (in alphabetical order)
+FEATURES_PROVIDED += periph_timer
+FEATURES_PROVIDED += periph_uart
```
Comment (gebart) [inline boards/pic32-wifire/Makefile.features]: Why is periph_uart in FEATURES_PROVIDED if there is no UART available?
according to comment at https://github.com/RIOT-OS/RIOT/pull/6066/files#diff-8790a8833e3d374cd14889b9f0e69a2cR18

Comment (neiljay) [inline boards/pic32-wifire/Makefile.features]: Uart Tx is implemented but not the full UART API yet.

Comment (neiljay) [inline boards/pic32-wifire/Makefile.features]: I tweaked the comment  when squashing see https://github.com/RIOT-OS/RIOT/pull/6066/commits/e2d7d3617ec1e7753d0a3e2406eebdf285d4563d

**Comment by neiljay** [comment]
@gebart @OlegHahm @haukepetersen Can we finally get this ACK'd and merged ? :-)

**Comment by gebart** [comment]
I don't have any of the boards to test it. I'm fine with the change as long as Murdock is fine.

**Comment by neiljay** [comment]
@gebart Cheers, can you ACK your change-request then. @OlegHahm has a board we donated.

**Comment by neiljay** [comment]
Please can someone merge this @OlegHahm @haukepetersen @miri64 @kaspar030 :-)

**Comment by kYc0o** [comment]
This code is very similar to #6092, so I'll have more or less the same comments. I'd like to have the comments addressed before merging this one too...

**Comment by neiljay** [comment]
Can some one re-run the CI Please, I don't believe the error.

**Comment by kaspar030** [comment]
> Can some one re-run the CI Please, I don't believe the error.

Murdock (not Murdock2) is our current last-word, so this PR basically passed CI.

**Comment by neiljay** [comment]
Cheers, Can we get this merged at last ?

File: boards/pic32-wifire/pic32_config_settings.c [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+volatile uint32_t DEVSIGN_LA  __attribute__ ((used, section(".devsign_la")))  = 0x7FFFFFFF;
+volatile uint32_t ADEVSIGN_LA __attribute__ ((used, section(".adevsign_la"))) = 0x7FFFFFFF;
+volatile uint32_t DEVSIGN_B1  __attribute__ ((used, section(".devsign_b1")))  = 0x7FFFFFFF;
+volatile uint32_t ADEVSIGN_B1 __attribute__ ((used, section(".adevsign_b1"))) = 0x7FFFFFFF;
+volatile uint32_t DEVSIGN_B2  __attribute__ ((used, section(".devsign_b2")))  = 0x7FFFFFFF;
+volatile uint32_t ADEVSIGN_B2 __attribute__ ((used, section(".adevsign_b2"))) = 0x7FFFFFFF;
+
+
+/*
+ * Without a reference to this function from elsewhere LD throws the whole
```
Comment (kaspar030) [inline boards/pic32-wifire/pic32_config_settings.c]: Probably has to do with missing '--whole-archive'. Could you please check if #5757 removes the need for this?

Comment (kaspar030) [inline boards/pic32-wifire/pic32_config_settings.c]: (well, let's not delay the PR because of this. can be checked at a later time.)

Comment (neiljay) [inline boards/pic32-wifire/pic32_config_settings.c]: Hi,

Yes that has helped cheers, when #5757 gets merged I'll happily remove this.

Neil

**Comment by neiljay** [comment]
Now squashed some more, Please Please Please can we finally get this merged.....

**Comment by haukepetersen** [comment]
Murdock is happy, and we got two ACKs -> lets go

@neiljay thanks a lot for all the efforts and your patience!

**Comment by neiljay** [comment]
Thanks everyone!!
Now for pic32mx...

**Pull request merged**